app/Http/Controllers/ElectionController.php

```diff
@@ -166,20 +170,25 @@ public function batchvote(Request $request, $election_id)
 
         $ranks = array();
 
-        // iterate over list of rankings
-        foreach($request->json()->get('votes') as $vote) {
-            // TODO: check this is a valid candidate?
-            $candidate_id = $vote['candidate_id'];
-            $rank_num = $vote['rank'];
-
-            $rank = CandidateRank::firstOrNew(['elector_id' => $elector->id, 'candidate_id' => $candidate_id]);
-            $rank->elector_id = $elector->id;
-            $rank->candidate_id = $candidate_id;
-            $rank->rank = $rank_num;
-            $rank->save();
+        // iterate over list of rankings and create/update candidate ranks in a transaction.
+        DB::transaction(function () use ($request, $elector){
```

Comment:
I think this line needs to be like this:

DB::transaction(function () use (&$ranks, $request, $elector){

Otherwise, $ranks is not defined inside the inner function.

Reply:
Yep totally right, missed that.